File: DesignPatternsExercises/java/TemplateMethodDesignPattern/Solution/Travel.java
// public __________(4)__________ class __________(5)__________ {
public abstract class Travel {
    protected String startDate, endDate, destination, meal, vehicle;
    protected boolean booked = false;

    // __________(6)__________ void __________(7)__________() {
    final void templateTravel() {
        System.out.println("Template travel information:");
        setDates();
        setDestination();
        if (!usePublicTransport())
            usePrivateVehicle();
        if (isMealAvailable())
            setMeal();
        if (usePublicTransport())
            bookTicket();
        displayTravelDetails();
    }

    abstract void setDates();

    abstract void setDestination();

    abstract void setMeal();

    abstract void usePrivateVehicle();

    abstract void bookTicket();

    // void __________(8)__________() {
    void displayTravelDetails() {
        System.out.println(" Travel has been confirmed.");
        System.out.println("    Start Date: " + startDate);
        System.out.println("    End Date: " + endDate);
        System.out.println("    Destination: " + destination);
    }

    boolean usePublicTransport() {
        return true;
    }

    // boolean __________(9)__________() {
    boolean isMealAvailable() {
        return false;
    }
}

File: DesignPatternsExercises/ruby/template_method_design_pattern/solution/travel.rb
class Travel
  # def __________(7)__________
  def template_travel
    puts 'Template travel information:'
    dates
    destination
    use_private_vehicle unless use_public_transport?
    meal if meal_available?
    book_ticket if use_public_transport?
    display_travel_details
  end

  def dates
    raise NotImplementedError, 'Subclasses must override this method'
  end

  def destination
    raise NotImplementedError, 'Subclasses must override this method'
  end

  def meal
    raise NotImplementedError, 'Subclasses must override this method'
  end

  def use_private_vehicle
    raise NotImplementedError, 'Subclasses must override this method'
  end

  def book_ticket
    raise NotImplementedError, 'Subclasses must override this method'
  end

  # def __________(8)__________
  def display_travel_details
    puts ' Travel has been confirmed.'
    puts "    Start Date: #{@start_date}"
    puts "    End Date: #{@end_date}"
    puts "    Destination: #{@destination}"
  end

  def use_public_transport?
    true
  end

  # def __________(9)__________
  def meal_available?
    false
  end
end
